atl-checker/src/algorithms/certain_zero/search_strategy/linear_constrained_phi.rs

```diff
@@ -0,0 +1,253 @@
+    /// Cached LinearConstrainedPhi of propositions. The bool is true for negated propositions.
+    proposition_cache: RefCell<HashMap<(Proposition, bool), LinearConstrainedPhi>>,
+}
+
+impl ConstrainedPhiMapper {
+    pub fn new() -> ConstrainedPhiMapper {
+        ConstrainedPhiMapper {
+            proposition_cache: RefCell::new(HashMap::new()),
+        }
+    }
+
+    /// Map the given phi to a linear constrained phi. Propositions will be map based on the given
+    /// game and they will be cached, so the same game should be used every time.
+    pub fn map(&self, game: &IntermediateLcgs, phi: &Phi) -> LinearConstrainedPhi {
+        self.map_phi_to_constraints(game, phi, false)
+    }
+
+    /// Takes a phi and maps it to a LinearConstrainedPhi
+    fn map_phi_to_constraints(
+        &self,
+        game: &IntermediateLcgs,
+        phi: &Phi,
+        negated: bool,
+    ) -> LinearConstrainedPhi {
+        match phi {
+            Phi::True => {
+                if !negated {
+                    LinearConstrainedPhi::True
+                } else {
+                    LinearConstrainedPhi::False
+                }
+            }
+            Phi::False => {
+                if !negated {
+                    LinearConstrainedPhi::False
+                } else {
+                    LinearConstrainedPhi::True
+                }
+            }
+            Phi::Proposition(proposition) => {
+                // If we get to a proposition, continue the mapping inside the proposition's condition
+                // The result may be cached
+                let key = (*proposition, negated);
+                let maybe_constraint = self.proposition_cache.borrow().get(&key).cloned();
+                maybe_constraint.unwrap_or_else(|| {
+                    // We have not mapped this proposition yet
+                    let decl = game.label_index_to_decl(*proposition);
+                    if let DeclKind::Label(label) = &decl.kind {
+                        let mapped_expr = self.map_expr_to_constraints(&label.condition, negated);
+                        // Save result in cache
+                        self.proposition_cache
+                            .borrow_mut()
+                            .insert(key, mapped_expr.clone());
+                        mapped_expr
+                    } else {
+                        panic!("Non-propositions symbol in ATL formula")
+                    }
+                })
+            }
+            Phi::Not(formula) => self.map_phi_to_constraints(game, formula, !negated),
+            Phi::Or(lhs, rhs) => {
+                let lhs_lcp = self.map_phi_to_constraints(game, lhs, negated);
+                let rhs_lcp = self.map_phi_to_constraints(game, rhs, negated);
+                if !negated {
+                    LinearConstrainedPhi::Or(Box::from(lhs_lcp), Box::from(rhs_lcp))
+                } else {
+                    LinearConstrainedPhi::And(Box::from(lhs_lcp), Box::from(rhs_lcp))
```

Comment:
DeMorgan?

Reply:
Yes. Also notice that `lhs_lcp` and `rhs_lcp` were negated if `negated` was true.

Reply:
Yes, I was about to make a comment about they should be negated in the new `or` and `and`, until I noticed this fact